### ISSUE TITLE
Can't build on NDK

### ISSUE THREAD

**Issue by sheosi** [DoumanAsh/opusic-sys#8]
Hi, I've trying to use this with Android NDK, however I've been running into problems, namely those:
`     ld: error: cannot open crtbegin_dynamic.o: No such file or directory
      ld: error: cannot open crtend_android.o: No such file or directory`

I'm using [](https://github.com/shekohex/flutterust) (which uses cargo-make and cargo-ndk) and I'm able to compile rust files and even some C FFI, I've even tried setting sysroot (though I've read is no recommended) but no luck.

The whole output is here: [](https://gist.github.com/sheosi/9b0fe8a980603a7981e489ca8704ef54)

Any ideas?

**Comment by DoumanAsh** [comment]
Can you output full log?
I suspect it is coming from libopus build itself, while it does use cmake, it is not necessary able to inherit NDK properties

**Comment by DoumanAsh** [comment]
First of all probably would need to propagate some CMAKE properties like:

```
-DCMAKE_ANDROID_ARCH_ABI=arm64-v8a -DBUILD_SHARED_LIBS=ON -DCMAKE_SYSTEM_NAME=Android -DCMAKE_SYSTEM_VERSION=21 -DCMAKE_ANDROID_NDK=<PATH_TO_SDK>
```

**Comment by DoumanAsh** [comment]
@bbqsrc I have question/request to you as author of `cargo-ndk`: is it possible to to export env variables that contains ABI and  API version?
While it works in general case, there are some tricky issues like propagating android vars to cmake and other tools

**Comment by bbqsrc** [comment]
@DoumanAsh sure, I accept pull requests. 😄

**Comment by DoumanAsh** [comment]
@sheosi I made first minimal solution that doesn't require cargo-ndk to re-export anything, this should be enough to let cmake know how to build using your NDK toolchain
https://github.com/DoumanAsh/opusic-sys/commit/878f6969a4886b8829f460768ad13a5361dba21d

Could you please to try?

**Comment by bbqsrc** [comment]
You may try the main branch of cargo-ndk now. It should export `CARGO_NDK_CMAKE_TOOLCHAIN_PATH` and `CARGO_NDK_ANDROID_TARGET`. Let me know if this suits your uses, and I'll release it with 2.3.0.

**Comment by DoumanAsh** [comment]
Thanks, I wanted first to determine what would be required.
Generally toolchain target and ABI is enough, but first we should let @sheosi verify it

**Comment by bbqsrc** [comment]
No problem, just wanted to check how easy it would be to _add_ these env vars at all to the current codebase. If any changes are needed to make it work, should be straightforward to do a PR now. 😄

**Comment by DoumanAsh** [comment]
@sheosi We can take `ANDROID_PLATFORM` from cargo-ndk https://github.com/bbqsrc/cargo-ndk/pull/20
But other variables will need to be provided by you. Important are: `ANDROID_STL`, `ANDROID_ARM_MODE`, `ANDROID_ARM_NEON`

**Comment by sheosi** [comment]
It's weird, what I get is:
`  thread 'main' panicked at 'a `libclang` shared library is not loaded on this thread', /home/sergio/.cargo/registry/src/github.com-1ecc6299db9ec823/clang-sys-0.29.3/src/lib.rs:1659:1`

[](https://gist.github.com/sheosi/48cfb38685bc0d3ec9426b863db82f86)

I got this before, but setting LIBCLANG_PATH or repeating the command made the compiler throw the other error. Anyway, I'd say this one is fixed. 

Btw, I'll keep those variables annotated.

**Comment by DoumanAsh** [comment]
Huh, I'm not sure where it is coming from.
Do you get it during build? Bindings are pre-generated so you don't need bindgen

**Comment by sheosi** [comment]
I get it during build, but from what I can gather seems to come from `dart-bindgen` [](https://github.com/sunshine-protocol/dart-bindgen), which is using a pretty old `clang-sys` I'll try to replace it with dart's `ffigen`-

**Comment by DoumanAsh** [comment]
@sheosi it should be fairly easy to get dart bindings yourself.
You can also consider using pre-generated `dart-sys` and take a look at build script to see how to generate bindings youtself

**Comment by DoumanAsh** [comment]
@sheosi btw I imagine you have similar use case as at my work, here is crate with bindings that I use https://crates.io/crates/dart-sdk-sys
It is pre-generated and should work out of box

**Comment by sheosi** [comment]
@DoumanAsh actually, I'm using Rust from Dart (making a Flutter app and calling Rust), I don't know how you would use `dart-sdk-sys` there. Basically is this: [](https://dev.to/sunshine-chain/dart-meets-rust-a-match-made-in-heaven-9f5).

I don't know if you can share anything, but are you using Dart from Rust?

**Comment by DoumanAsh** [comment]
We basically use Dart isolate to allow call Rust code from Dart, but there is also communication back which is done via Dart_Port so not sure if you'd need it indeed if you just call Rust code without needing to call Dart back

**Comment by sheosi** [comment]
Oh, that's clever. I don't need it right now, but I might in the future.

**Comment by DoumanAsh** [comment]
@sheosi do you have NDK build fixed now with master?

**Comment by sheosi** [comment]
Yeah, it is compiling just fine. 😄️

**Comment by DoumanAsh** [comment]
Ok good, please do not silently close issue next time, I want to be sure it is fixed.
Published 0.3.5 with these changes
